File: Android/app/src/main/java/com/auralis/crisisconnect/messaging/AuthorityMlsMessagePayload.kt
package com.auralis.crisisconnect.messaging

import java.nio.ByteBuffer
import java.nio.charset.CodingErrorAction
import java.nio.charset.StandardCharsets
import java.util.Base64
import org.json.JSONArray
import org.json.JSONObject

data class AuthorityMlsMessagePayload(
    val recipientUid: String,
    val recipientName: String,
    val senderName: String,
    val text: String,
    val sentAtMillis: Long,
    val attachments: List<ChannelAttachment> = emptyList(),
    val replyToId: String? = null,
)

/** Strict, cross-platform AuthorityChat application envelope carried only as MLS plaintext. */
object AuthorityMlsMessagePayloadCodec {
    private const val MAX_PAYLOAD_BYTES = 900_000
    private const val MAX_TEXT_BYTES = 64 * 1024
    private const val MAX_ATTACHMENTS = 8
    private val ID = Regex("^[A-Za-z0-9_-]{1,128}$")
    private val PATH = Regex(
        "^authorityMessageAttachments/am2_[A-Za-z0-9_-]{43}/[^/]{1,256}/" +
            "[0-9A-Fa-f]{8}-[0-9A-Fa-f]{4}-4[0-9A-Fa-f]{3}-[89AaBb][0-9A-Fa-f]{3}-[0-9A-Fa-f]{12}$"
    )

    fun encode(payload: AuthorityMlsMessagePayload): ByteArray {
        validate(payload)
        val root = JSONObject()
            .put("version", 2)
            .put("kind", "message")
            .put("recipientUid", payload.recipientUid)
            .put("recipientName", payload.recipientName)
            .put("senderName", payload.senderName)
            .put("text", payload.text)
            .put("sentAtMillis", payload.sentAtMillis)
        val descriptors = JSONArray()
        payload.attachments.forEach { attachment ->
            val descriptor = JSONObject()
                .put("path", attachment.path)
                .put("nonce", attachment.nonce)
                .put("key", attachment.key)
                .put("name", attachment.name)
                .put("mime", attachment.mime)
                .put("size", attachment.size)
            attachment.width?.let { descriptor.put("width", it) }
            attachment.height?.let { descriptor.put("height", it) }
            attachment.durationSec?.let { descriptor.put("duration", it) }
            descriptors.put(descriptor)
        }
        root.put("attachments", descriptors)
        payload.replyToId?.let { root.put("replyToId", it) }
        return root.toString().toByteArray(StandardCharsets.UTF_8).also {
            require(it.size in 1..MAX_PAYLOAD_BYTES) { "Authority MLS message payload is too large." }
        }
    }

    fun decode(encoded: ByteArray): AuthorityMlsMessagePayload {
        require(encoded.size in 1..MAX_PAYLOAD_BYTES) { "Authority MLS message payload size is invalid." }
        val text = StandardCharsets.UTF_8.newDecoder()
            .onMalformedInput(CodingErrorAction.REPORT)
            .onUnmappableCharacter(CodingErrorAction.REPORT)
            .decode(ByteBuffer.wrap(encoded)).toString()
        val root = JSONObject(text)
        require(root.optInt("version", -1) == 2 && root.optString("kind") == "message") {
            "Authority MLS message payload version is invalid."
        }
        val attachments = root.optJSONArray("attachments") ?: throw SecurityException("Attachment list is missing.")
        require(attachments.length() <= MAX_ATTACHMENTS) { "Too many Authority MLS attachments." }
        val decodedAttachments = buildList {
            for (index in 0 until attachments.length()) add(decodeAttachment(attachments.getJSONObject(index)))
        }
        return AuthorityMlsMessagePayload(
            recipientUid = bounded(strictString(root, "recipientUid"), 256, allowEmpty = false),
            recipientName = bounded(strictString(root, "recipientName"), 256, allowEmpty = true),
            senderName = bounded(strictString(root, "senderName"), 256, allowEmpty = true),
            text = bounded(strictString(root, "text"), MAX_TEXT_BYTES, allowEmpty = true),
            sentAtMillis = strictLong(root, "sentAtMillis"),
            attachments = decodedAttachments,
            replyToId = if (root.has("replyToId")) strictString(root, "replyToId") else null,
        ).also(::validate)
    }

    private fun validate(payload: AuthorityMlsMessagePayload) {
        bounded(payload.recipientUid, 256, allowEmpty = false)
        bounded(payload.recipientName, 256, allowEmpty = true)
        bounded(payload.senderName, 256, allowEmpty = true)
        bounded(payload.text, MAX_TEXT_BYTES, allowEmpty = true)
        require(payload.sentAtMillis > 0 && payload.attachments.size <= MAX_ATTACHMENTS) {
            "Authority MLS message timestamp or attachment count is invalid."
        }
        require(payload.text.isNotEmpty() || payload.attachments.isNotEmpty()) { "Authority MLS message is empty." }
        payload.attachments.forEach(::validateAttachment)
        require(payload.replyToId == null || ID.matches(payload.replyToId)) { "Authority MLS reply ID is invalid." }
    }

    private fun decodeAttachment(raw: JSONObject): ChannelAttachment = ChannelAttachment(
        path = strictString(raw, "path"),
        nonce = strictString(raw, "nonce"),
        key = strictString(raw, "key"),
        name = strictString(raw, "name"),
        mime = strictString(raw, "mime"),
        size = strictLong(raw, "size"),
        width = if (raw.has("width")) strictInt(raw, "width") else null,
        height = if (raw.has("height")) strictInt(raw, "height") else null,
        durationSec = if (raw.has("duration")) strictInt(raw, "duration") else null,
    ).also(::validateAttachment)

    private fun validateAttachment(attachment: ChannelAttachment) {
        val key = attachment.key ?: throw SecurityException("Authority MLS attachment key is missing.")
        require(PATH.matches(attachment.path) && decodeCanonicalBase64(attachment.nonce).size == 12 &&
            decodeCanonicalBase64(key).size == 32 && attachment.name.toByteArray().size in 1..255 &&
            attachment.mime.toByteArray().size in 1..255 && attachment.size in 0..ChannelAttachments.MAX_ATTACHMENT_BYTES) {
            "Authority MLS attachment descriptor is invalid."
        }
    }

    private fun bounded(value: String, maxBytes: Int, allowEmpty: Boolean): String {
        val size = value.toByteArray(StandardCharsets.UTF_8).size
        require((allowEmpty || value.isNotEmpty()) && size <= maxBytes &&
            value.none { it.code in 0..8 || it.code in 11..12 || it.code in 14..31 || it.code == 127 }) {
            "Authority MLS message field is invalid."
        }
        return value
    }

    private fun decodeCanonicalBase64(value: String): ByteArray {
        val decoded = Base64.getDecoder().decode(value)
        require(Base64.getEncoder().encodeToString(decoded) == value) { "Non-canonical base64." }
        return decoded
    }

    private fun strictString(root: JSONObject, name: String): String =
        (root.get(name) as? String) ?: throw SecurityException("Authority MLS $name must be a string.")

    private fun strictLong(root: JSONObject, name: String): Long {
        val number = root.get(name) as? Number ?: throw SecurityException("Authority MLS $name must be numeric.")
        val value = number.toLong()
        require(number.toDouble().isFinite() && number.toDouble() == value.toDouble()) {
            "Authority MLS $name must be an integer."
        }
        return value
    }

    private fun strictInt(root: JSONObject, name: String): Int {
        val value = strictLong(root, name)
        require(value in 0..1_000_000) { "Authority MLS $name is out of range." }
        return value.toInt()
    }
}

File: iOS/Crisis Connect/Features/Messaging/AuthorityMlsMessagePayload.swift
import Foundation

struct AuthorityMlsMessagePayload: Sendable {
    let recipientUid: String
    let recipientName: String
    let senderName: String
    let text: String
    let sentAtMillis: Int64
    let attachments: [ChannelAttachment]
    let replyToId: String?
}

enum AuthorityMlsMessagePayloadError: Error {
    case malformed
    case tooLarge
}

/// Strict cross-platform AuthorityChat application envelope carried only inside MLS plaintext.
enum AuthorityMlsMessagePayloadCodec {
    private static let maxPayloadBytes = 900_000
    private static let maxTextBytes = 64 * 1024
    private static let maxAttachments = 8
    private static let idPattern = "^[A-Za-z0-9_-]{1,128}$"
    private static let pathPattern = "^authorityMessageAttachments/am2_[A-Za-z0-9_-]{43}/[^/]{1,256}/[0-9A-Fa-f]{8}-[0-9A-Fa-f]{4}-4[0-9A-Fa-f]{3}-[89AaBb][0-9A-Fa-f]{3}-[0-9A-Fa-f]{12}$"

    private struct WirePayload: Codable {
        let version: Int
        let kind: String
        let recipientUid: String
        let recipientName: String
        let senderName: String
        let text: String
        let sentAtMillis: Int64
        let attachments: [WireAttachment]
        let replyToId: String?
    }

    private struct WireAttachment: Codable {
        let path: String
        let nonce: String
        let key: String?
        let name: String
        let mime: String
        let size: Int
        let width: Int?
        let height: Int?
        let duration: Int?
    }

    static func encode(_ payload: AuthorityMlsMessagePayload) throws -> Data {
        try validate(payload)
        let encoded = try JSONEncoder().encode(WirePayload(
            version: 2,
            kind: "message",
            recipientUid: payload.recipientUid,
            recipientName: payload.recipientName,
            senderName: payload.senderName,
            text: payload.text,
            sentAtMillis: payload.sentAtMillis,
            attachments: payload.attachments.map {
                WireAttachment(
                    path: $0.path,
                    nonce: $0.nonce,
                    key: $0.keyBase64,
                    name: $0.name,
                    mime: $0.mime,
                    size: $0.size,
                    width: $0.width,
                    height: $0.height,
                    duration: $0.durationSec
                )
            },
            replyToId: payload.replyToId
        ))
        guard !encoded.isEmpty, encoded.count <= maxPayloadBytes else { throw AuthorityMlsMessagePayloadError.tooLarge }
        return encoded
    }

    static func decode(_ encoded: Data) throws -> AuthorityMlsMessagePayload {
        guard !encoded.isEmpty, encoded.count <= maxPayloadBytes else { throw AuthorityMlsMessagePayloadError.tooLarge }
        let raw = try JSONDecoder().decode(WirePayload.self, from: encoded)
        guard raw.version == 2, raw.kind == "message" else { throw AuthorityMlsMessagePayloadError.malformed }
        let payload = AuthorityMlsMessagePayload(
            recipientUid: raw.recipientUid,
            recipientName: raw.recipientName,
            senderName: raw.senderName,
            text: raw.text,
            sentAtMillis: raw.sentAtMillis,
            attachments: raw.attachments.map {
                ChannelAttachment(
                    path: $0.path,
                    nonce: $0.nonce,
                    keyBase64: $0.key,
                    name: $0.name,
                    mime: $0.mime,
                    size: $0.size,
                    width: $0.width,
                    height: $0.height,
                    durationSec: $0.duration
                )
            },
            replyToId: raw.replyToId
        )
        try validate(payload)
        return payload
    }

    private static func validate(_ payload: AuthorityMlsMessagePayload) throws {
        try validateString(payload.recipientUid, maxBytes: 256, allowEmpty: false)
        try validateString(payload.recipientName, maxBytes: 256, allowEmpty: true)
        try validateString(payload.senderName, maxBytes: 256, allowEmpty: true)
        try validateString(payload.text, maxBytes: maxTextBytes, allowEmpty: true)
        guard payload.sentAtMillis > 0,
              payload.attachments.count <= maxAttachments,
              !payload.text.isEmpty || !payload.attachments.isEmpty,
              payload.replyToId == nil || payload.replyToId!.range(of: idPattern, options: .regularExpression) != nil else {
            throw AuthorityMlsMessagePayloadError.malformed
        }
        try payload.attachments.forEach(validateAttachment)
    }

    private static func validateAttachment(_ attachment: ChannelAttachment) throws {
        guard attachment.path.range(of: pathPattern, options: .regularExpression) != nil,
              let nonce = canonicalBase64(attachment.nonce), nonce.count == 12,
              let keyValue = attachment.keyBase64,
              let key = canonicalBase64(keyValue), key.count == 32,
              !attachment.name.isEmpty, attachment.name.utf8.count <= 255,
              !attachment.mime.isEmpty, attachment.mime.utf8.count <= 255,
              (0...ChannelAttachments.maxAttachmentBytes).contains(attachment.size) else {
            throw AuthorityMlsMessagePayloadError.malformed
        }
    }

    private static func validateString(_ value: String, maxBytes: Int, allowEmpty: Bool) throws {
        guard (allowEmpty || !value.isEmpty), value.utf8.count <= maxBytes,
              value.unicodeScalars.allSatisfy({ scalar in
                  let code = scalar.value
                  return !(code <= 8 || code == 11 || code == 12 || (14...31).contains(code) || code == 127)
              }) else { throw AuthorityMlsMessagePayloadError.malformed }
    }

    private static func canonicalBase64(_ value: String) -> Data? {
        guard let decoded = Data(base64Encoded: value), decoded.base64EncodedString() == value else { return nil }
        return decoded
    }
}
